openlr/openlr_match_quality/openlr_assessment_tool/mainwindow.cpp

```diff
@@ -38,50 +49,184 @@ class TrafficDrawerDelegate : public TrafficDrawerDelegateBase
-    CHECK(!sample.GetItems().empty(), ("Sample must not be empty."));
-    auto const & points = sample.GetPoints(sampleIndex);
+    CHECK(!points.empty(), ("Points must not be empty."));
 
     LOG(LINFO, ("Decoded segment", points));
-    m_drapeApi.AddLine(NextLineId(),
+    m_drapeApi.AddLine(kDecodedLineId,
                        df::DrapeApiLineData(points, dp::Color(0, 0, 255, 255))
                        .Width(3.0f).ShowPoints(true /* markPoints */));
   }
 
-  void DrawEncodedSegment(openlr::LinearSegment const & segment) override
+  void DrawEncodedSegment(std::vector<m2::PointD> const & points) override
   {
-    auto const & points = segment.GetMercatorPoints();
-
     LOG(LINFO, ("Encoded segment", points));
-    m_drapeApi.AddLine(NextLineId(),
+    m_drapeApi.AddLine(kEncodedLineId,
                        df::DrapeApiLineData(points, dp::Color(255, 0, 0, 255))
                        .Width(3.0f).ShowPoints(true /* markPoints */));
   }
 
-  void Clear() override
+  void DrawGoldenPath(std::vector<m2::PointD> const & points) override
+  {
+    m_drapeApi.AddLine(kGoldenLineId,
+                       df::DrapeApiLineData(points, dp::Color(255, 127, 36, 255))
+                       .Width(4.0f).ShowPoints(true /* markPoints */));
+  }
+
+  void ClearGoldenPath() override
+  {
+    m_drapeApi.RemoveLine(kGoldenLineId);
+  }
+
+  void ClearAllPaths() override
   {
     m_drapeApi.Clear();
   }
 
-private:
-  string NextLineId() { return strings::to_string(m_lineId++); }
 
-  uint32_t m_lineId = 0;
+  void VisualizePoints(std::vector<m2::PointD> const & points) override
+  {
+    UserMarkControllerGuard g(m_bm, UserMarkType::DEBUG_MARK);
+    g.m_controller.SetIsVisible(true);
+    g.m_controller.SetIsDrawable(true);
+    for (auto const & p : points)
+      g.m_controller.CreateUserMark(p);
+  }
+
+  void ClearAllVisualizedPoints() override
+  {
+    UserMarkControllerGuard g(m_bm, UserMarkType::DEBUG_MARK);
+    g.m_controller.Clear();
+  }
 
+private:
   Framework & m_framework;
   df::DrapeApi & m_drapeApi;
+  BookmarkManager & m_bm;
+};
+
+bool PointsMatch(m2::PointD const & a, m2::PointD const & b)
+{
+  auto constexpr kToleraneDistanceM = 1.0;
+  return MercatorBounds::DistanceOnEarth(a, b) < kToleraneDistanceM;
+}
+
+class PointsControllerDelegate : public PointsControllerDelegateBase
+{
+public:
+  PointsControllerDelegate(Framework & framework)
+    : m_framework(framework)
+    , m_index(framework.GetIndex())
+    , m_roadGraph(m_index, routing::IRoadGraph::Mode::ObeyOnewayTag,
+                  make_unique<routing::CarModelFactory>())
+  {
+  }
+
+  std::vector<m2::PointD> GetAllJunctionPointsInViewPort() const
+  {
+    std::vector<m2::PointD> points;
+    auto const & rect = m_framework.GetCurrentViewport();
+    auto const pushPoint = [&points, &rect](m2::PointD const & point) {
+      if (!rect.IsPointInside(point))
+        return;
+      for (auto const & p : points)
+      {
+        if (PointsMatch(point, p))
```

Comment:
Even better! You may modify the old code for the case when number of points is small, and kill two rabbits simultaneously!